<a href="https://colab.research.google.com/github/Aayush-Rathod-11/Netflix_EDA/blob/main/Netflix_EDA_Project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


# 1. Problem Statement
You are a Data Analyst hired at a company. Your manager has given you a real-world dataset and asked you
to understand the data, find patterns and trends, visualize the key insights, and write a clear report with 5
business or research recommendations based on your findings.

# 2. Import Libraries

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

print("All libraries imported successfully!")

All libraries imported successfully!


# 3. Load Dataset

In [ ]:
import pandas as pd

data = pd.read_csv('/content/drive/My Drive/AI_ML_Internship/Project_1_Netflix_EDA/dataset/netflix.csv')

# 4. Data Inspection

In [ ]:
# Check dataset shap

print("=" * 50)
print("DATASET OVERVIEW")
print("=" * 50)
print(f"Shape: {data.shape[0]} rows × {data.shape[1]} columns\n")

DATASET OVERVIEW
Shape: 8807 rows × 12 columns



In [ ]:
# Check column names

print("COLUMN NAMES & DATA TYPES:")
print(data.dtypes)
print()

COLUMN NAMES & DATA TYPES:
show_id         object
type            object
title           object
director        object
cast            object
country         object
date_added      object
release_year     int64
rating          object
duration        object
listed_in       object
description     object
dtype: object



In [ ]:
# Display first 5 rows

print("FIRST 5 ROWS:")
display(data.head())

FIRST 5 ROWS:


,show_id,type,title,director,cast,country,date_added,release_year,rating,duration,listed_in,description
0,s1,Movie,Dick Johnson Is Dead,Kirsten Johnson,NaN,United States,"September 25, 2021",2020,PG-13,90 min,Documentaries,"As her father nears the end of his life, filmm..."
1,s2,TV Show,Blood & Water,NaN,"Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...",South Africa,"September 24, 2021",2021,TV-MA,2 Seasons,"International TV Shows, TV Dramas, TV Mysteries","After crossing paths at a party, a Cape Town t..."
2,s3,TV Show,Ganglands,Julien Leclercq,"Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...",NaN,"September 24, 2021",2021,TV-MA,1 Season,"Crime TV Shows, International TV Shows, TV Act...",To protect his family from a powerful drug lor...
3,s4,TV Show,Jailbirds New Orleans,NaN,NaN,NaN,"September 24, 2021",2021,TV-MA,1 Season,"Docuseries, Reality TV","Feuds, flirtations and toilet talk go down amo..."
4,s5,TV Show,Kota Factory,NaN,"Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...",India,"September 24, 2021",2021,TV-MA,2 Seasons,"International TV Shows, Romantic TV Shows, TV ...",In a city of coaching centers known to train I...


In [ ]:
# Statistical Summary

print("\nBASIC STATISTICS:")
display(data.describe(include='all'))


BASIC STATISTICS:


,show_id,type,title,director,cast,country,date_added,release_year,rating,duration,listed_in,description
count,8807,8807,8807,6173,7982,7976,8797,8807.000000,8803,8804,8807,8807
unique,8807,2,8804,4528,7692,748,1767,NaN,17,220,514,8775
top,s8807,Movie,Feb-09,Rajiv Chilaka,David Attenborough,United States,"January 1, 2020",NaN,TV-MA,1 Season,"Dramas, International Movies","Paranormal activity at a lush, abandoned prope..."
freq,1,6131,2,19,19,2818,109,NaN,3207,1793,362,4
mean,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2014.180198,NaN,NaN,NaN,NaN
std,NaN,NaN,NaN,NaN,NaN,NaN,NaN,8.819312,NaN,NaN,NaN,NaN
min,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1925.000000,NaN,NaN,NaN,NaN
25%,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2013.000000,NaN,NaN,NaN,NaN
50%,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2017.000000,NaN,NaN,NaN,NaN
75%,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2019.000000,NaN,NaN,NaN,NaN


In [ ]:
# Find Missing Values

print("\nMISSING VALUES PER COLUMN:")
missing = data.isnull().sum()
missing_pct = (missing / len(data) * 100).round(2)
missing_data = pd.DataFrame({'Missing Count': missing, 'Missing %': missing_pct})
print(missing_data[missing_data['Missing Count'] > 0])


MISSING VALUES PER COLUMN:
            Missing Count  Missing %
director             2634      29.91
cast                  825       9.37
country               831       9.44
date_added             10       0.11
rating                  4       0.05
duration                3       0.03


In [ ]:
# Find Duplicate Records

print("\nDUPLICATE ROWS:", data.duplicated().sum())


DUPLICATE ROWS: 0


* The Netflix dataset contains **8,807 rows and 12 columns**, representing movies and TV shows available on Netflix globally.

* Key columns include: type (Movie/TV Show), title, director,
country, release_year, rating, and listed_in (genre).

* The dataset has missing values in director (30%), cast (9.5%),
and country (6.5%) columns.

* There are no duplicate rows. Most columns are object (string) type;
release_year and duration need type conversion before analysis.

* Data spans from 1925 to 2021, with the majority added after 2015.



# 5. Data Cleaning

In [ ]:
# Initialize data_clean for data cleaning operations
df_clean = data.copy()

# Remove duplicates
before = len(df_clean)
df_clean.drop_duplicates(inplace=True)
print(f"Duplicates removed: {before - len(df_clean)}")

Duplicates removed: 0


In [ ]:
# 2. Fill missing 'director' with 'Unknown'
#    Reason: dropping rows would lose 30% of data.
#    'Unknown' is a valid category for analysis.
df_clean['director'].fillna('Unknown', inplace=True)

/tmp/ipykernel_1766/1203342671.py:4: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_clean['director'].fillna('Unknown', inplace=True)


In [ ]:
# 3. Fill missing 'country' with 'Unknown'
#    Reason: same as director — too many missing to drop.
df_clean['country'].fillna('Unknown', inplace=True)

/tmp/ipykernel_1766/1892088012.py:3: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_clean['country'].fillna('Unknown', inplace=True)


In [ ]:
# 4. Drop rows where 'date_added' is missing
#    Reason: only 10 rows affected; we need this for time analysis.
df_clean.dropna(subset=['date_added'], inplace=True)

In [ ]:
# 5. Convert 'date_added' to datetime
#    Reason: stored as string; need datetime for year/month extraction.

# Ensure 'date_added' is string type before stripping, to handle re-execution
df_clean['date_added'] = df_clean['date_added'].astype(str)

df_clean['date_added'] = pd.to_datetime(
    df_clean['date_added'].str.strip(), errors='coerce', format='mixed'
)
df_clean['year_added'] = df_clean['date_added'].dt.year
df_clean['month_added'] = df_clean['date_added'].dt.month

In [ ]:
# 6. Drop irrelevant columns
#    Reason: 'cast' has 9.5% missing and won't be used in analysis.

# Check if 'cast' column exists before dropping
if 'cast' in df_clean.columns:
    df_clean.drop(columns=['cast'], inplace=True)

print(f"\nRows before cleaning: {len(data)}")
print(f"Rows after cleaning:  {len(df_clean)}")
print(f"\nRemaining missing values:")
print(df_clean.isnull().sum()[df_clean.isnull().sum() > 0])
print("\nData types after cleaning:")
print(df_clean.dtypes)


Rows before cleaning: 8807
Rows after cleaning:  8797

Remaining missing values:
rating      4
duration    3
dtype: int64

Data types after cleaning:
show_id                 object
type                    object
title                   object
director                object
country                 object
date_added      datetime64[ns]
release_year             int64
rating                  object
duration                object
listed_in               object
description             object
year_added               int32
month_added              int32
dtype: object


In [ ]:
df_clean['rating'].fillna(df_clean['rating'].mode()[0], inplace=True)

/tmp/ipykernel_1766/1648633727.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_clean['rating'].fillna(df_clean['rating'].mode()[0], inplace=True)


In [ ]:
# Drop rows where 'duration' is missing
initial_rows = len(df_clean)
df_clean.dropna(subset=['duration'], inplace=True)
print(f"Rows dropped due to missing duration: {initial_rows - len(df_clean)}")

# Verify no more missing 'duration' values
print("\nRemaining missing values after handling duration:")
print(df_clean.isnull().sum()[df_clean.isnull().sum() > 0])

Rows dropped due to missing duration: 3

Remaining missing values after handling duration:
Series([], dtype: int64)


In [ ]:
# 7. Verify cleaning
print("Shape after cleaning:", df_clean.shape)
print("Missing values after cleaning:\n", df_clean.isnull().sum())
print("Data types after cleaning:\n", df_clean.dtypes)

Shape after cleaning: (8794, 13)
Missing values after cleaning:
 show_id         0
type            0
title           0
director        0
country         0
date_added      0
release_year    0
rating          0
duration        0
listed_in       0
description     0
year_added      0
month_added     0
dtype: int64
Data types after cleaning:
 show_id                 object
type                    object
title                   object
director                object
country                 object
date_added      datetime64[ns]
release_year             int64
rating                  object
duration                object
listed_in               object
description             object
year_added               int32
month_added              int32
dtype: object


- Duplicate rows were removed to avoid repeated records.
- Missing values in director, cast, and country columns were replaced with meaningful labels.
- Missing ratings were filled using the mode value because it represents the most common category.
- The date_added column was converted into datetime format for easier analysis.
- After cleaning, the dataset became more consistent and suitable for exploratory data analysis.

# 6. Exploratory Data Analysis

##### Q1: Content type distribution
##### Q2: Top 10 content-producing countries
##### Q3: Which month sees the most content additions?
##### Q4: Content added per year
##### Q5: Which year has the highest number of releases?

In [ ]:
print("Q1: Content type distribution")
print("=" * 40)
type_counts = df_clean['type'].value_counts()
type_pct = df_clean['type'].value_counts(normalize=True) * 100
print(type_counts)
print("\nPercentage:")
print(type_pct.round(2))

Q1: Content type distribution
type
Movie      6128
TV Show    2666
Name: count, dtype: int64

Percentage:
type
Movie      69.68
TV Show    30.32
Name: proportion, dtype: float64


In [ ]:
print("\nQ2: Top 10 content-producing countries")
print("=" * 40)
top_countries = (df_clean[df_clean['country'] != 'Unknown']
                 ['country']
                 .str.split(',')
                 .explode()
                 .str.strip()
                 .value_counts()
                 .head(10))
print(top_countries)


Q2: Top 10 content-producing countries
country
United States     3681
India             1046
United Kingdom     805
Canada             445
France             393
Japan              317
Spain              232
South Korea        231
Germany            226
Mexico             169
Name: count, dtype: int64


In [ ]:
print("\nQ3: Which month sees the most content additions?")
print("=" * 40)
df_clean['month_added'] = df_clean['date_added'].dt.month_name()
df_clean['month_added'].value_counts()


Q3: Which month sees the most content additions?


,count
month_added,
July,827
December,813
September,769
April,763
October,760
August,754
March,742
January,738
June,728


In [ ]:
print("\nQ4: Content added per year")
print("=" * 40)
yearly = df_clean.groupby('year_added').size().reset_index(name='count')
print(yearly.sort_values('year_added'))


Q4: Content added per year
    year_added  count
0         2008      2
1         2009      2
2         2010      1
3         2011     13
4         2012      3
5         2013     11
6         2014     24
7         2015     82
8         2016    427
9         2017   1187
10        2018   1649
11        2019   2016
12        2020   1879
13        2021   1498


In [ ]:
print("\nQ5: Which year has the highest number of releases?")
print("=" * 40)
df_clean['release_year'].value_counts().head(10)
# OR
# df_clean.groupby('release_year').size().sort_values(ascending=False).head(10)


Q5: Which year has the highest number of releases?


,count
release_year,
2018,1146
2017,1031
2019,1030
2020,953
2016,901
2021,592
2015,557
2014,352
2013,287


- Movies are significantly more numerous than TV Shows, indicating that Netflix focuses heavily on movie content.

# 7. Visualizations

# 8. Insights Report

# 9. Conclusion